server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -204,13 +202,18 @@ public String toString() {
         });
     }
 
-    private void onPrimaryDemoted(Exception demotionFailure) {
-        String primaryFail = String.format(Locale.ROOT,
-            "primary shard [%s] was demoted while failing replica shard",
-            primary.routingEntry());
-        // we are no longer the primary, fail ourselves and start over
-        primary.failShard(primaryFail, demotionFailure);
-        finishAsFailed(new RetryOnPrimaryException(primary.routingEntry().shardId(), primaryFail, demotionFailure));
+    private void onNoLongerPrimary(Exception failure) {
+        final String message;
+        if (failure instanceof ShardStateAction.NoLongerPrimaryShardException) {
+            // we are no longer the primary, fail ourselves and start over
+            message = String.format(Locale.ROOT, "primary shard [%s] was demoted while failing replica shard", primary.routingEntry());
+            primary.failShard(message, failure);
+        } else {
+            // these can occur if the node is shutting down and are okay any other exception here is not expected and merits investigation.
+            assert failure instanceof NodeClosedException || failure instanceof TransportException : failure;
```

Comment:
@bleskes mentioned that the TransportException here should be coming from https://github.com/elastic/elasticsearch/blob/022726011ce1d9e4748fd39293a3aeae6954b3ac/server/src/main/java/org/elasticsearch/transport/TransportService.java#L645-L649

We should verify this here with an assertion and (I think in a follow-up) look into throwing a more appropriate exception in TransportService. Possible options are `AlreadyClosedException` or a custom subclass of `TransportException`.

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/39584/commits/6afb094f04d21f2f30c601a4beea9fa25de06c6c.